src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -125389,14 +125541,110 @@ public void CompareExchange_LocationNullState(string location, string value, str
+            comp.VerifyDiagnostics(isLocationNonNullAfterCall
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (19,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(19, 13)
+                });
 
+            source = $@"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{{
+    public static class Interlocked
+    {{
+        public static T CompareExchange<T>(ref T location, T value, T comparand) where T : class? {{ return location; }}
+    }}
+}}
+
+class C
+{{
+    // locationNonNull is annotated, but its flow state is non-null
+    void M<T>(T? locationNonNull, T? locationMaybeNull, T comparandNonNull, T? comparandMaybeNull, T valueNonNull, T? valueMaybeNull) where T : class, new()
+    {{
+        locationNonNull = new T();
+        Interlocked.CompareExchange(ref {location}, {value}, {comparand});
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNullAfterCall
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (19,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(19, 13)
+                });
+
+            source = $@"
+class C
+{{
+    // locationNonNull is annotated, but its flow state is non-null
+    void M(object? locationNonNull, object? locationMaybeNull, object comparandNonNull, object? comparandMaybeNull, object valueNonNull, object? valueMaybeNull)
+    {{
+        locationNonNull = new object();
+        if ({location} == {comparand})
+        {{
+            {location} = {value};
+        }}
+        _ = {location}.ToString();
+    }}
+}}
+    ";
+            comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(isLocationNonNullAfterCall
+                ? Array.Empty<DiagnosticDescription>()
+                : new[]
+                {
+                    // (12,13): warning CS8602: Dereference of a possibly null reference.
+                    //         _ = {location}.ToString();
+                    Diagnostic(ErrorCode.WRN_NullReferenceReceiver, location).WithLocation(12, 13)
+                });
+        }
+
+        // This test should be merged back into CompareExchange_LocationNullState when the issue with inference with null is resolved
+        // Currently differs by a diagnostic
+        [Theory]
+        [InlineData("locationNonNull", "null", "comparandNonNull", false)]
```

Comment:
Should we also add `[InlineData("locationNonNull", "null", "null", false)]` here?